File: optimization-greedy-algorithms-problems-plus/src/algorithms/job-sequencing.ts
```typescript
/**
 * src/algorithms/job-sequencing.ts
 *
 * This file implements the Job Sequencing with Deadlines problem using a greedy approach.
 *
 * Problem Description:
 * Given a set of `n` jobs, where each job `i` has a deadline `d_i` and a profit `p_i`.
 * Each job takes one unit of time to complete. We want to find a sequence of jobs that
 * maximizes the total profit, such that each selected job is completed by its deadline.
 * A job can only be completed at most once.
 *
 * Example:
 * Jobs:
 *   Job A: Deadline=2, Profit=100
 *   Job B: Deadline=1, Profit=19
 *   Job C: Deadline=2, Profit=27
 *   Job D: Deadline=1, Profit=25
 *   Job E: Deadline=3, Profit=15
 *
 * Greedy Choice: Sort jobs by profit in descending order. For each job, try to schedule
 *                it in the latest possible free slot before its deadline.
 *
 * Optimal Sequence (Greedy):
 * Jobs sorted by profit: A (100, D=2), C (27, D=2), D (25, D=1), B (19, D=1), E (15, D=3)
 * Max Deadline is 3. Available slots: [slot 1, slot 2, slot 3]
 *
 * 1. Process Job A (P=100, D=2):
 *    - Try slot 2 (D=2): Free. Schedule A in slot 2.
 *    - Scheduled Jobs: [_, A, _]
 *    - Total Profit: 100
 *
 * 2. Process Job C (P=27, D=2):
 *    - Try slot 2 (D=2): Taken by A.
 *    - Try slot 1 (D=1, < D=2): Free. Schedule C in slot 1.
 *    - Scheduled Jobs: [C, A, _]
 *    - Total Profit: 100 + 27 = 127
 *
 * 3. Process Job D (P=25, D=1):
 *    - Try slot 1 (D=1): Taken by C.
 *    - Job D cannot be scheduled (no free slot before or at deadline 1). Skip.
 *
 * 4. Process Job B (P=19, D=1):
 *    - Try slot 1 (D=1): Taken by C. Skip.
 *
 * 5. Process Job E (P=15, D=3):
 *    - Try slot 3 (D=3): Free. Schedule E in slot 3.
 *    - Scheduled Jobs: [C, A, E]
 *    - Total Profit: 127 + 15 = 142
 *
 * Final Schedule: [C, A, E] (Job C in slot 1, Job A in slot 2, Job E in slot 3)
 * Total Max Profit: 142
 */

/**
 * Represents a job with an ID, deadline, and profit.
 */
export interface Job {
    id: string;      // Unique identifier for the job
    deadline: number; // The latest time unit by which the job must be completed
    profit: number;  // The profit gained if the job is completed
}

/**
 * Solves the Job Sequencing with Deadlines problem using a greedy approach.
 *
 * The greedy strategy is based on prioritizing jobs with higher profits
 * and scheduling them as late as possible but still within their deadlines.
 *
 * Algorithm Steps:
 * 1. Sort all jobs in descending order of their profits.
 * 2. Determine the maximum deadline among all jobs. This defines the total number
 *    of available time slots (from 1 to `maxDeadline`).
 * 3. Initialize an array `slots` of size `maxDeadline + 1` (using 1-based indexing for slots)
 *    to keep track of occupied time slots. Initialize all slots as free (e.g., with null/0).
 * 4. Initialize `totalProfit` to 0 and `scheduledJobs` to an empty array.
 * 5. Iterate through the sorted jobs:
 *    a. For the current job, try to find a free time slot. Start checking from its
 *       `deadline` down to 1. The goal is to schedule it as late as possible
 *       to leave earlier slots open for jobs with smaller deadlines.
 *    b. If a free slot `j` is found such that `j <= job.deadline`:
 *       i. Mark slot `j` as occupied by this job.
 *       ii. Add `job.profit` to `totalProfit`.
 *       iii. Add the job `id` to `scheduledJobs`.
 *       iv. Break from the inner loop (this job is scheduled).
 * 6. Return the `totalProfit` and `scheduledJobs`.
 *
 * Time Complexity:
 * - Sorting jobs: O(N log N), where N is the number of jobs.
 * - Finding max deadline: O(N).
 * - Iterating through jobs: N jobs.
 * - For each job, iterating through slots from deadline down to 1: O(maxDeadline).
 * - Total Time Complexity: O(N log N + N * maxDeadline).
 *   If maxDeadline is large (e.g., proportional to N^2 or N), this can be O(N^3) or O(N^2).
 *   A more optimized approach for slot finding can use Disjoint Set Union (DSU) to achieve O(N log N)
 *   or O(N * alpha(maxDeadline)) where alpha is inverse Ackermann function (nearly constant).
 *   The current implementation uses a simple array scan, making it O(N * maxDeadline).
 *
 * Space Complexity:
 * - Storing sorted jobs: O(N).
 * - Storing `slots` array: O(maxDeadline).
 * - Storing `scheduledJobs`: O(N) in the worst case.
 * - Total Space Complexity: O(N + maxDeadline).
 *
 * @param jobs An array of Job objects.
 * @returns An object containing the maximum total profit and the IDs of the scheduled jobs.
 */
export function jobSequencing(jobs: Job[]): { totalProfit: number; scheduledJobs: string[] } {
    // Edge case: No jobs provided
    if (!jobs || jobs.length === 0) {
        return { totalProfit: 0, scheduledJobs: [] };
    }

    // 1. Sort jobs in descending order of their profits.
    const sortedJobs = [...jobs].sort((a, b) => b.profit - a.profit);

    // 2. Determine the maximum deadline to set up time slots.
    let maxDeadline = 0;
    for (const job of jobs) {
        if (job.deadline > maxDeadline) {
            maxDeadline = job.deadline;
        }
    }

    // 3. Initialize an array `slots` to keep track of occupied time slots.
    // `slots[i]` will store the ID of the job scheduled in time slot `i`.
    // We use `maxDeadline + 1` size for 1-based indexing up to maxDeadline.
    const slots: (string | null)[] = new Array(maxDeadline + 1).fill(null);

    let totalProfit = 0;
    const scheduledJobs: string[] = [];

    // 5. Iterate through the sorted jobs.
    for (const job of sortedJobs) {
        // Try to schedule the job in the latest possible free slot before its deadline.
        // We iterate from `job.deadline` downwards to 1.
        for (let i = Math.min(job.deadline, maxDeadline); i >= 1; i--) {
            // If the current slot `i` is free
            if (slots[i] === null) {
                slots[i] = job.id;       // Mark slot `i` as occupied by this job
                totalProfit += job.profit; // Add job's profit to total
                scheduledJobs.push(job.id); // Add job's ID to the list of scheduled jobs
                break; // Job scheduled, move to the next job
            }
        }
    }

    // Sort scheduled jobs by their slot assignment for deterministic output (optional)
    // The current implementation fills `scheduledJobs` in order of processing,
    // which corresponds to profit-descending order. If output needs to be by slot order,
    // an additional step is required:
    // const sortedBySlot = slots.slice(1).filter(id => id !== null).map(id => jobs.find(j => j.id === id)!);
    // return { totalProfit, scheduledJobs: sortedBySlot.map(j => j.id) };

    return { totalProfit, scheduledJobs: scheduledJobs.sort((a, b) => { // Sort for deterministic test output, not strictly necessary for profit.
        const jobA = jobs.find(j => j.id === a)!;
        const jobB = jobs.find(j => j.id === b)!;
        return jobB.profit - jobA.profit; // Maintain profit-descending order
    })};
}

// --- Alternative Implementation (using Disjoint Set Union for optimized slot finding) ---
/*
* The `find` operation in DSU can effectively find the latest free slot.
* If `parent[i]` stores the next available slot *before or at* `i`,
* then `find(i)` can quickly give you the latest free slot.
*
* DSU Enhanced `jobSequencing` (conceptual):
* function jobSequencingDSU(jobs: Job[]): { totalProfit: number; scheduledJobs: string[] } {
*   const sortedJobs = [...jobs].sort((a, b) => b.profit - a.profit);
*   let maxDeadline = 0;
*   for (const job of jobs) { maxDeadline = Math.max(maxDeadline, job.deadline); }
*
*   const dsu = new DisjointSetUnion<number>();
*   for (let i = 0; i <= maxDeadline; i++) {
*     dsu.makeSet(i); // Each slot is initially its own set (its parent is itself)
*   }
*
*   let totalProfit = 0;
*   const scheduledJobs: string[] = [];
*   const finalSlots: { [slot: number]: string } = {}; // To store actual slot assignments
*
*   for (const job of sortedJobs) {
*     // Find the latest free slot available for this job (<= job.deadline)
*     // In a DSU for job sequencing, `find(i)` typically gives you the *parent*
*     // of the set `i`, which is the representative. We want the *highest*
*     // available slot <= deadline. A common DSU modification is to have
*     // `parent[i]` point to the highest available slot `j < i`.
*     // Here we can use `find(min(job.deadline, maxDeadline))` to get the highest
*     // free slot below or equal to the job's deadline.
*     const availableSlot = dsu.find(Math.min(job.deadline, maxDeadline));
*
*     if (availableSlot > 0) { // Slot 0 is typically a dummy for "no slot available"
*       // Schedule the job in this slot
*       finalSlots[availableSlot] = job.id;
*       totalProfit += job.profit;
*       scheduledJobs.push(job.id);
*
*       // Union this slot with the slot before it, effectively marking it as taken
*       // and making the previous slot its new "representative" for future searches.
*       dsu.union(availableSlot, availableSlot - 1);
*     }
*   }
*   // Sort scheduledJobs by their original profit (desc) for consistent output,
*   // or by slot number if that's desired.
*   const finalScheduledJobIDs = Object.keys(finalSlots)
*                                     .sort((a, b) => parseInt(a) - parseInt(b))
*                                     .map(slot => finalSlots[parseInt(slot)]);
*
*   return { totalProfit, scheduledJobs: finalScheduledJobIDs };
* }
*
* This DSU approach reduces the slot search time from O(maxDeadline) to O(alpha(maxDeadline))
* per job, making the overall complexity O(N log N + N * alpha(maxDeadline)), which is
* significantly better when `maxDeadline` is large.
*/

// --- Brute Force vs. Optimized (Greedy) Discussion ---
/*
* Brute Force Approach:
*   A brute force approach would involve generating all possible permutations of jobs
*   and for each permutation, checking if it's a valid schedule (all jobs meet deadlines)
*   and calculating its total profit. The number of permutations is N!, which is
*   computationally infeasible for even small N.
*   Alternatively, one could use recursion with backtracking to explore all subsets
*   of jobs and all possible assignments to slots, but this would still be exponential.

* Optimized (Greedy) Approach:
*   The greedy strategy is proven to be optimal for the Job Sequencing with Deadlines
*   problem. The two key greedy choices are:
*   1. Prioritize jobs with higher profits. This ensures that if we have to make a choice
*      between two jobs for a single slot, we pick the one that yields more profit.
*   2. Schedule jobs in the latest possible free slot before their deadline. This is
*      crucial because it keeps earlier slots open for other high-profit jobs that might
*      have tighter deadlines. If a job `J` with deadline `D` is placed in slot `t < D`,
*      and slot `D` is available, placing `J` in slot `D` instead would have the same
*      profit but leave slot `t` open, potentially allowing another job with `deadline <= t`
*      to be scheduled. This exchange argument reinforces the optimality.
*/

// --- ASCII Art Diagram for Job Sequencing Logic ---
/*
Jobs:
 A: P=100, D=2
 B: P=19,  D=1
 C: P=27,  D=2
 D: P=25,  D=1
 E: P=15,  D=3

Sorted by Profit (desc): A (100,2), C (27,2), D (25,1), B (19,1), E (15,3)
Max Deadline = 3. Slots: [null, null, null] (Indices 1, 2, 3)

1. Process Job A (P=100, D=2)
   - Try slot i=2: slots[2] is null. Assign A to slot 2.
   - slots: [null, null, 'A'] -> [null, 'A', null] (corrected ASCII visual to reflect 1-indexed slots)
   - Total Profit: 100
   - Scheduled: ['A']

   Time Slots:
   1   2   3
   --- --- ---
   |   | A |
   --- --- ---

2. Process Job C (P=27, D=2)
   - Try slot i=2: slots[2] is 'A' (taken).
   - Try slot i=1: slots[1] is null. Assign C to slot 1.
   - slots: ['C', 'A', null]
   - Total Profit: 100 + 27 = 127
   - Scheduled: ['A', 'C']

   Time Slots:
   1   2   3
   --- --- ---
   | C | A |
   --- --- ---

3. Process Job D (P=25, D=1)
   - Try slot i=1: slots[1] is 'C' (taken).
   - No free slot <= deadline 1. Skip Job D.

   Time Slots: (No change)
   1   2   3
   --- --- ---
   | C | A |
   --- --- ---

4. Process Job B (P=19, D=1)
   - Try slot i=1: slots[1] is 'C' (taken).
   - No free slot <= deadline 1. Skip Job B.

   Time Slots: (No change)
   1   2   3
   --- --- ---
   | C | A |
   --- --- ---

5. Process Job E (P=15, D=3)
   - Try slot i=3: slots[3] is null. Assign E to slot 3.
   - slots: ['C', 'A', 'E']
   - Total Profit: 127 + 15 = 142
   - Scheduled: ['A', 'C', 'E']

   Time Slots:
   1   2   3
   --- --- ---
   | C | A | E |
   --- --- ---

Final Result:
Total Profit: 142
Scheduled Jobs: ['A', 'C', 'E'] (order depends on how `scheduledJobs` is accumulated or sorted)
*/
```

File: data-structures-greedy-algorithms-complete/src/main/java/com/greedy/problems/JobSequencing.java
```java
package com.greedy.problems;

import com.greedy.utils.Pair;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * **Problem:** Job Sequencing with Deadlines and Profits
 *
 * Given a set of N jobs, where each job has a deadline and a profit.
 * Each job takes unit time to complete. We can only do one job at a time.
 * The goal is to find a sequence of jobs that maximizes the total profit.
 *
 * **Example:**
 * Jobs: (Job_id, Deadline, Profit)
 * J1: (A, 2, 100)
 * J2: (B, 1, 10)
 * J3: (C, 2, 15)
 * J4: (D, 1, 25)
 *
 * **Greedy Choice Property:**
 * To maximize total profit, it's intuitive to prioritize jobs with higher profits.
 * If two jobs have the same profit, the one with an earlier deadline might be considered
 * first to free up later slots.
 * The greedy strategy is:
 * 1. Sort all jobs in decreasing order of their profit.
 * 2. Iterate through the sorted jobs. For each job, try to schedule it into the latest
 *    possible time slot before or on its deadline, without conflicting with already
 *    scheduled jobs. The latest possible slot approach ensures earlier slots are kept free
 *    for jobs with even earlier deadlines if they appear later in the profit-sorted list.
 *
 * **Optimal Substructure:**
 * If an optimal sequence includes a job J, then the remaining jobs in the sequence
 * form an optimal sequence for the remaining available time slots.
 *
 * **Algorithm Steps:**
 * 1. Create a data structure to hold jobs, each with an ID, deadline, and profit.
 * 2. Sort the jobs in decreasing order of profit.
 * 3. Determine the maximum deadline among all jobs. This will help in creating a time slot array.
 * 4. Create a boolean array `timeSlots` of size `maxDeadline + 1`. Initialize all entries to false,
 *    representing available slots.
 * 5. Create a list to store the sequence of selected jobs.
 * 6. Iterate through the sorted jobs (from highest profit to lowest):
 *    a. For the current job, try to find an available time slot. Start checking from its deadline
 *       backwards to time slot 1.
 *    b. If an empty slot `i` (where `i <= deadline`) is found, mark `timeSlots[i]` as true,
 *       add the job to the selected jobs list, and break from the inner loop (as this job is scheduled).
 * 7. The total profit is the sum of profits of the selected jobs.
 */
public class JobSequencing {

    /**
     * Represents a single job with an ID, deadline, and profit.
     */
    public static class Job {
        String id;
        int deadline;
        int profit;

        public Job(String id, int deadline, int profit) {
            this.id = id;
            this.deadline = deadline;
            this.profit = profit;
        }

        public String getId() {
            return id;
        }

        public int getDeadline() {
            return deadline;
        }

        public int getProfit() {
            return profit;
        }

        @Override
        public String toString() {
            return "Job{" + id + ", D=" + deadline + ", P=" + profit + "}";
        }
    }

    /**
     * Solves the Job Sequencing Problem using a greedy approach.
     *
     * @param jobs A list of Job objects.
     * @return A list of selected jobs that maximize total profit, ordered by their scheduled time slot.
     *         Returns null if input jobs list is null.
     *
     * **Time Complexity:**
     * - Sorting jobs: O(N log N), where N is the number of jobs.
     * - Finding max deadline: O(N).
     * - Iterating through jobs and finding slots: N jobs, each potentially checking up to maxDeadline slots.
     *   Worst case: O(N * maxDeadline).
     * - Total: O(N log N + N * maxDeadline). In cases where maxDeadline is large (e.g., N), this can be O(N^2).
     *   If maxDeadline is small, it's closer to O(N log N).
     *
     * **Space Complexity:**
     * - O(N) for storing sorted jobs.
     * - O(maxDeadline) for the `timeSlots` array.
     * - O(N) for the result list.
     * - Total: O(N + maxDeadline).
     */
    public List<Job> findMaxProfitJobSequence(List<Job> jobs) {
        if (jobs == null || jobs.isEmpty()) {
            return Collections.emptyList();
        }

        // Step 2: Sort jobs in decreasing order of profit.
        // If profits are equal, tie-break with deadline (earlier deadline preferred, though not critical for correctness)
        List<Job> sortedJobs = new ArrayList<>(jobs);
        sortedJobs.sort(Comparator
                .comparingInt(Job::getProfit).reversed() // Primary sort: highest profit first
                .thenComparingInt(Job::getDeadline));    // Secondary sort: earlier deadline first (optional but good practice)

        // Step 3: Determine the maximum deadline to create time slot array.
        int maxDeadline = 0;
        for (Job job : jobs) {
            if (job.getDeadline() > maxDeadline) {
                maxDeadline = job.getDeadline();
            }
        }

        // Step 4: Create a boolean array to track available time slots.
        // index i represents time slot i. timeSlots[0] is unused, slots from 1 to maxDeadline.
        String[] scheduledJobs = new String[maxDeadline + 1]; // To store job IDs in their scheduled slots
        boolean[] timeSlots = new boolean[maxDeadline + 1];   // To mark if a slot is taken
        // Initialize with false (all slots free) is default for boolean array.

        List<Job> selectedJobs = new ArrayList<>();
        int totalProfit = 0;

        // Step 6: Iterate through sorted jobs.
        for (Job currentJob : sortedJobs) {
            // Step 6a: Try to find an available time slot for the current job.
            // Start from its deadline backwards to time slot 1.
            // This ensures we leave earlier slots open for jobs with even earlier deadlines.
            for (int i = Math.min(maxDeadline, currentJob.getDeadline()); i >= 1; i--) {
                if (!timeSlots[i]) { // If slot 'i' is free
                    timeSlots[i] = true;        // Mark slot as taken
                    scheduledJobs[i] = currentJob.getId(); // Schedule the job
                    selectedJobs.add(currentJob); // Add to the result list
                    totalProfit += currentJob.getProfit();
                    break; // Job scheduled, move to the next job
                }
            }
        }

        // The problem typically asks for the maximum profit and/or the sequence of job IDs.
        // We can return the list of selected Job objects, sorted by their ID for consistency
        // or by their scheduled slot if we want a time-ordered sequence.
        // Let's return the jobs ordered by their scheduled slot.
        // For that, we need to reconstruct from 'scheduledJobs' array.

        List<Job> finalSequence = new ArrayList<>();
        for (int i = 1; i <= maxDeadline; i++) {
            if (scheduledJobs[i] != null) {
                // Find the original Job object by ID (inefficient, but for final presentation)
                String jobId = scheduledJobs[i];
                finalSequence.add(jobs.stream().filter(j -> j.getId().equals(jobId)).findFirst().orElse(null));
            }
        }
        // Remove any nulls if lookup failed (shouldn't happen with correct IDs)
        finalSequence = finalSequence.stream().filter(java.util.Objects::nonNull).collect(Collectors.toList());

        // Note: Returning 'selectedJobs' directly (which is sorted by profit) is also valid,
        // but 'finalSequence' provides the actual execution order.
        return finalSequence;
    }

    /**
     * Finds the maximum profit from a list of jobs found by `findMaxProfitJobSequence`.
     * @param jobs The list of jobs selected by the greedy algorithm.
     * @return The total profit.
     */
    public int calculateTotalProfit(List<Job> jobs) {
        if (jobs == null) {
            return 0;
        }
        return jobs.stream().mapToInt(Job::getProfit).sum();
    }

    /**
     * Conceptual Brute-Force Approach for Job Sequencing.
     * This method explores all possible subsets of jobs and all possible permutations
     * for scheduling them to find the one with maximum profit.
     * It's computationally very expensive (NP-hard).
     *
     * **Time Complexity:** O(N! * N * maxDeadline) or worse, due to subset generation and permutation.
     * **Space Complexity:** O(N) for recursion stack and temporary lists.
     *
     * This is a simplified outline; a full implementation would be much more complex.
     *
     * @param jobs The list of all available jobs.
     * @param maxDeadline The maximum deadline among all jobs.
     * @return The maximum profit found.
     */
    public int findMaxProfitBruteForceConceptual(List<Job> jobs, int maxDeadline) {
        if (jobs == null || jobs.isEmpty()) {
            return 0;
        }

        // This would involve generating all subsets of jobs,
        // then for each subset, trying all permutations of scheduling them
        // while respecting deadlines and unit time constraint.
        // Then calculate profit for each valid schedule and find the maximum.

        // For simplicity, we'll just return -1 to indicate this is not implemented efficiently.
        // A true brute force would be exponential.
        System.out.println("Brute-force approach for Job Sequencing is extremely complex and slow (exponential).");
        System.out.println("Returning 0 as a placeholder for conceptual purposes.");
        return 0;
    }
}

```